src/models/release-asset.ts

```diff
@@ -0,0 +1,444 @@
+) {
+	const assetHelpers = once(
+		() =>
+			(opts.isBrowser
+				? // eslint-disable-next-line @typescript-eslint/no-require-imports
+					(require('../util/asset-helpers.browser') as typeof import('../util/asset-helpers.browser'))
+				: // eslint-disable-next-line @typescript-eslint/no-require-imports
+					(require('../util/asset-helpers') as typeof import('../util/asset-helpers'))
+			).assetHelpers,
+	);
+	const { pine, request } = deps;
+
+	const parseUrlForRequest = (href: string) => {
+		const url = new URL(href);
+		const baseUrl = `${url.protocol}//${url.host}`;
+		const pathname = url.pathname;
+		const qs: Record<string, string> = {};
+
+		url.searchParams.forEach((value, key) => {
+			qs[key] = value;
+		});
+
+		return {
+			baseUrl,
+			url: pathname,
+			qs,
+			sendToken: false,
+		};
+	};
+
+	const uploadPart = async (
+		asset: File | string,
+		part: UploadPart,
+		requestedPartSize: number,
+	) => {
+		let body: Blob | ArrayBuffer;
+		const offset = (part.partNumber - 1) * requestedPartSize;
+
+		if (typeof asset === 'string') {
+			const buffer = await assetHelpers().readFileChunk(
+				asset,
+				offset,
+				part.chunkSize,
+			);
+			body = buffer.buffer.slice(
+				buffer.byteOffset,
+				buffer.byteOffset + buffer.byteLength,
+			);
+		} else {
+			const end = offset + part.chunkSize;
+			body = asset.slice(offset, end);
+		}
+
+		// We directly use fetch API rather than balena-request because
+		// pre-signed S3 URLs contain authentication in query parameters
+		// and must be used exactly as provided without modification (encoding/decoding for balena-request)
+		const res = await fetch(part.url, {
+			method: 'PUT',
+			body,
+		});
+
+		const ETag = res.headers.get('ETag')?.replace(/^"+|"+$/g, '');
+		if (ETag == null || typeof ETag !== 'string') {
+			throw new Error(`Error on the received ETag ${ETag}`);
+		}
+		return {
+			PartNumber: part.partNumber,
+			ETag,
+		};
+	};
+
+	const uploadParts = async (
+		asset: File | string,
+		parts: UploadPart[],
+		uploadParams: Required<Pick<UploadParams, 'chunkSize' | 'parallelUploads'>>,
+		totalSize: number,
+		onUploadProgress?: (progress: UploadProgressInfo) => void | Promise<void>,
+	) => {
+		let totalUploaded = 0;
+		return await limitedMap(
+			parts,
+			async (part) => {
+				const result = await uploadPart(asset, part, uploadParams.chunkSize);
+				totalUploaded += part.chunkSize;
+				if (onUploadProgress) {
+					await onUploadProgress({
+						chunkLength: part.chunkSize,
+						totalUploaded: Math.min(totalUploaded, totalSize),
+						totalToUpload: totalSize,
+					});
+				}
+				return result;
+			},
+			{ concurrency: uploadParams.parallelUploads },
+		);
+	};
+
+	const uploadMultipartReleaseAsset = async (
+		releaseAssetId: number,
+		asset: File | string,
+		uploadParams: Required<Pick<UploadParams, 'chunkSize' | 'parallelUploads'>>,
+		onUploadProgress?: (progress: UploadProgressInfo) => void | Promise<void>,
```

Comment:
NIT: since this is only called internally and we want the sdk to pass that parameter in all calls
```suggestion
		onUploadProgress: undefined | (progress: UploadProgressInfo) => (void | Promise<void>),
```